File: Leetcode/SetMatrixZeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int n = matrix.size();
        if(n == 0) return;
        int m = matrix[0].size();
        int r[n + 1], c[m + 1];
        for(int i = 0; i <= n; ++i) r[i] = 0;
        for(int j = 0; j <= m; ++j) c[j] = 0;
        for(int i = 0; i < n; ++i){
            for(int j = 0; j < m; ++j) {
                if(matrix[i][j] == 0) r[i] = 1, c[j] = 1;
            }
        }
        for(int i = 0; i < n; ++i){
            if(r[i]==1){
                for(int j = 0; j < m; ++j){
                    matrix[i][j] = 0;
                }
            }    
            else{
                for(int j = 0; j < m; ++j){
                    if(c[j]==1) matrix[i][j] = 0;
                }
            }
        }
    }
};

File: Leetcode/SetMatrixZeroes.rs
// https://leetcode.com/problems/set-matrix-zeroes

struct Solution {}

impl Solution {
    pub fn set_zeroes(matrix: &mut Vec<Vec<i32>>) {
        let n = matrix.len();
        let m = matrix[0].len();

        let mut x = 0;
        loop {
            let mut found = false;
            'out: for i in 0..n {
                for j in 0..m {
                    if matrix[i][j] == x {
                        found = true;
                        break 'out;
                    }
                }
            }

            if !found {
                break;
            } else {
                x += 1;
            }
        }

        for i in 0..n {
            for j in 0..m {
                if matrix[i][j] == 0 {
                    continue;
                }
                for i1 in 0..n {
                    if matrix[i1][j] == 0 {
                        matrix[i][j] = x;
                        break;
                    }
                }
                for j1 in 0..m {
                    if matrix[i][j1] == 0 {
                        matrix[i][j] = x;
                        break;
                    }
                }
            }
        }

        for i in 0..n {
            for j in 0..m {
                if matrix[i][j] == x {
                    matrix[i][j] = 0;
                }
            }
        }
    }
}
